### **Week 7 Assignment**
#### Sandhya Mainali
#### Presidential Graduate School
#### PRG 330: Python Programming
#### Professor Pant
#### Apr 20,2025

### Section A: Algorithm and Data Structure Optimization

#### Task 1

In [1]:
import timeit
import random

# Version 1: Brute-force approach
def find_pairs_brute_force(nums, target):
    pairs = []
    n = len(nums)
    for i in range(n):
        for j in range(i + 1, n):
            if nums[i] + nums[j] == target:
                pairs.append((nums[i], nums[j]))
    return pairs

# Function to test performance with different input sizes
def test_brute_force_performance():
    input_sizes = [1000, 5000, 10000]
    target = 100
    results = []

    for size in input_sizes:
        nums = [random.randint(0, 100) for _ in range(size)]
        time_taken = timeit.timeit(
            stmt=lambda: find_pairs_brute_force(nums, target),
            number=1
        )
        results.append((size, time_taken))
    
    return results

# Run and print results
def print_results(results):
    print(f"{'Input Size':<12} | {'Execution Time (s)':<20}")
    print("-" * 35)
    for size, time_taken in results:
        print(f"{size:<12} | {time_taken:<20.6f}")

# Main driver
if __name__ == "__main__":
    results = test_brute_force_performance()
    print("Brute-Force Version Performance:\n")
    print_results(results)


Brute-Force Version Performance:

Input Size   | Execution Time (s)  
-----------------------------------
1000         | 0.063160            
5000         | 1.143734            
10000        | 3.937583            


- O(n²)
- The reason for using nested loops involves iterating over succeeding elements for all current elements.
- Apart from the result list the algorithm exclusively operates through i, j and other scalar variables.
- Extremely poor with big datasets.
- Applications that demand real-time performance should avoid the usage of this method.

In [2]:
import timeit
import random
from IPython.display import display

# Version 2: Optimized approach using a set
def find_pairs_optimized(nums, target):
    seen = set()
    pairs = set()

    for num in nums:
        complement = target - num
        if complement in seen:
            pairs.add(tuple(sorted((num, complement))))
        seen.add(num)

    return list(pairs)

# Function to test performance with different input sizes
def test_optimized_performance():
    input_sizes = [1000, 10000, 100000]
    target = 100
    results = []

    for size in input_sizes:
        nums = [random.randint(0, 100) for _ in range(size)]
        time_taken = timeit.timeit(
            stmt=lambda: find_pairs_optimized(nums, target),
            number=1
        )
        results.append((size, time_taken))
    
    return results

# Function to print results in table format
def fun(results):
    print(f"{'Input Size':<12} | {'Execution Time (s)':<20}")
    print("-" * 35)
    for size, time_taken in results:
        display(f"{size:<12} | {time_taken:<20.6f}")

# Main driver
if __name__ == "__main__":
    results = test_optimized_performance()
    display("Optimized Version Performance:\n")
    fun(results)


'Optimized Version Performance:\n'

Input Size   | Execution Time (s)  
-----------------------------------


'1000         | 0.000734            '

'10000        | 0.008181            '

'100000       | 0.047230            '

- The time complexity is O(n) according to this version which presents the information well.
- The program functions proficiently with more than 100,000 elements in its system.
- Every element receives one visit.
- The operation duration for sets lasts an average of O(1).
- Each pair addition to the set takes the same duration as sorting because pair length remains constant at 2.

### Section B: Code Profiling and Memory Management 

#### Task 2


### Befor Optimizatio using cprofile

In [12]:
# Generate a sample text block with 1500+ words
sample_text = (
    "data science machine learning artificial intelligence " * 200 +
    "python pandas numpy matplotlib seaborn sklearn " * 150 +
    "model training validation accuracy precision recall f1-score " * 100
)

# Save to 'file.txt'
with open("file.txt", "w", encoding='utf-8') as f:
    f.write(sample_text)

print("Created 'file.txt' with over 1500 words.")

Created 'file.txt' with over 1500 words.


In [13]:
import cProfile
from collections import Counter
import re
import os

def word_frequency(file_path):
    """
    Reads a large text file and returns the frequency of each word in descending order.

    Args:
        file_path (str): Path to the text file.

    Returns:
        List[Tuple[str, int]]: List of (word, frequency) tuples.
    """
    if not os.path.exists(file_path):
        raise FileNotFoundError(f"File not found: {file_path}")
    
    with open(file_path, 'r', encoding='utf-8') as file:
        text = file.read()
    
    words = re.findall(r'\b\w+\b', text.lower())
    word_counts = Counter(words)
    return word_counts.most_common()

def profile_word_frequency(file_path, top_n=10):
    """
    Profiles the word frequency function and prints the top `n` words.

    Args:
        file_path (str): Path to the text file.
        top_n (int): Number of top words to display.
    """
    try:
        result = word_frequency(file_path)
        print(f"\nTop {top_n} most frequent words:\n")
        for word, count in result[:top_n]:
            print(f"{word}: {count}")
    except FileNotFoundError as e:
        print(e)

# Set your file path here
file_path = 'file.txt'  # Modify this path based on your file's location

# Run profiling
cProfile.run('profile_word_frequency(file_path)', sort='cumtime')


Top 10 most frequent words:

data: 200
science: 200
machine: 200
learning: 200
artificial: 200
intelligence: 200
python: 150
pandas: 150
numpy: 150
matplotlib: 150
         1045 function calls (1030 primitive calls) in 0.035 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      2/1    0.000    0.000    0.034    0.034 {built-in method builtins.exec}
      2/1    0.001    0.000    0.034    0.034 <string>:1(<module>)
      2/1    0.000    0.000    0.034    0.034 2596923764.py:26(profile_word_frequency)
      2/1    0.000    0.000    0.021    0.021 decorator.py:229(fun)
        1    0.000    0.000    0.021    0.021 history.py:55(only_when_enabled)
        1    0.000    0.000    0.019    0.019 history.py:845(writeout_cache)
        1    0.000    0.000    0.018    0.018 history.py:833(_writeout_input_cache)
        2    0.014    0.007    0.014    0.007 {method '__exit__' of 'sqlite3.Connection' objects}
        1    0.000    0

- Before optimization 1045 function run  in 0.035second

### After Optimization using cprofile

In [14]:
import cProfile
from collections import Counter
import heapq
import os

def word_frequency(file_path, top_n=None):
    """
    Reads a large text file and returns the frequency of each word in descending order.

    Args:
        file_path (str): Path to the text file.
        top_n (int, optional): Return only the top N most frequent words.

    Returns:
        List[Tuple[str, int]]: List of (word, frequency) tuples.
    """
    word_counts = Counter()

    # Read the file line by line for memory efficiency
    with open(file_path, 'r', encoding='utf-8') as file:
        for line in file:
            words = line.lower().split()  # Basic whitespace tokenizer
            word_counts.update(words)

    # Efficient top-N retrieval
    if top_n:
        return heapq.nlargest(top_n, word_counts.items(), key=lambda x: x[1])
    else:
        return word_counts.most_common()

def profile_word_frequency():
    file_path = "file.txt"  # Update this path as needed

    if not os.path.exists(file_path):
        print(f"File not found: {file_path}")
        return

    result = word_frequency(file_path, top_n=10)
    print("\nTop 10 most frequent words:\n")
    for word, freq in result:
        print(f"{word}: {freq}")

# Run profiling (works in scripts or Jupyter with minor tweaks)
if __name__ == "__main__":
    cProfile.run('profile_word_frequency()', sort='cumtime')


Top 10 most frequent words:

data: 200
science: 200
machine: 200
learning: 200
artificial: 200
intelligence: 200
python: 150
pandas: 150
numpy: 150
matplotlib: 150
         693 function calls (685 primitive calls) in 0.006 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.002    0.002 3629223846.py:31(profile_word_frequency)
        3    0.000    0.000    0.002    0.001 events.py:86(_run)
        3    0.000    0.000    0.002    0.001 {method 'run' of '_contextvars.Context' objects}
        1    0.000    0.000    0.002    0.002 3629223846.py:6(word_frequency)
      2/1    0.000    0.000    0.001    0.001 base_events.py:1909(_run_once)
        1    0.000    0.000    0.001    0.001 kernelbase.py:294(poll_control_queue)
        1    0.000    0.000    0.001    0.001 _base.py:537(set_result)
        1    0.000    0.000    0.001    0.001 _base.py:337(_invoke_callbacks)
        1    0.000    0.000 

- After optimization 693 function run in 0.006 second

## Before Optimization memory_profiler

In [15]:
def compute_word_frequencies(file_path):
    print("[*] compute_word_frequencies started")
    with open(file_path, 'r') as file:
        text = file.read()
    print("[*] File read successfully")
    words = text.split()
    frequencies = {}
    for word in words:
        word = word.lower()
        frequencies[word] = frequencies.get(word, 0) + 1
    print("[*] Word frequency computed")
    return frequencies

def run_frequency_analysis():
    print("[*] Running frequency analysis for memory profiling...")
    compute_word_frequencies("file.txt")

In [16]:
import os
print("[*] Current working directory:", os.getcwd())
print("[*] file.txt exists:", os.path.exists("file.txt"))

[*] Current working directory: c:\Users\sandh\OneDrive\Desktop\python
[*] file.txt exists: True


## After Optimizagion of memory-profiler 

In [18]:
from memory_profiler import profile

@profile
def word_frequency(file_path, top_n=None):
    """
    Reads a large text file and returns the frequency of each word in descending order.
    
    Args:
    - file_path (str): The path to the text file to read.
    - top_n (int): Limit the results to the top N most frequent words (optional).
    
    Returns:
    - List of tuples: A list of tuples where each tuple contains a word and its frequency, sorted by frequency in descending order.
    """
    word_counts = Counter()

    # Open and read the file line by line to reduce memory usage
    with open(file_path, 'r') as file:
        for line in file:
            # Convert the line to lowercase and split by spaces (words)
            words = line.lower().split()  # Simple split by spaces; works efficiently
            word_counts.update(words)  # Update counts with words from the line

    # If top_n is provided, get the top N most frequent words using heapq for efficiency
    if top_n:
        # Use heapq to get the top N frequent words in O(n log k) time complexity
        sorted_word_counts = heapq.nlargest(top_n, word_counts.items(), key=lambda x: x[1])
    else:
        # Otherwise, return all words sorted by frequency in descending order
        sorted_word_counts = word_counts.most_common()

    return sorted_word_counts

# Profiling the function using cProfile
def profile_word_frequency():
    file_path = r'file.txt'  # Ensure the file exists
    word_frequencies = word_frequency(file_path, top_n=10)
    print(word_frequencies)  # <--- Add this line


# Profile the function
if __name__ == "__main__":
    profile_word_frequency()  # This will show memory profiling output

ERROR: Could not find file C:\Users\sandh\AppData\Local\Temp\ipykernel_16924\3617867365.py
[('data', 200), ('science', 200), ('machine', 200), ('learning', 200), ('artificial', 200), ('intelligence', 200), ('python', 150), ('pandas', 150), ('numpy', 150), ('matplotlib', 150)]


## Section C: Concurrency and Built-in Libraries 

### Task 3

## version 1: Sequential (requests)

In [29]:
### version1
import requests
import time

urls = [
    "https://www.google.com/",
    "https://www.youtube.com/",
    "https://www.python.org",
    "https://www.wikipedia.org",
    "https://www.github.com"
]

def fetch_url(url):
    response = requests.get(url)
    return url, response.status_code

def sequential_download():
    start = time.time()
    results = [fetch_url(url) for url in urls]
    duration = time.time() - start
    for url, status in results:
        print(f"{url} -> {status}")
    print(f"Sequential time: {duration:.2f} seconds")

if __name__ == "__main__":
    sequential_download()


https://www.google.com/ -> 200
https://www.youtube.com/ -> 200
https://www.python.org -> 200
https://www.wikipedia.org -> 200
https://www.github.com -> 200
Sequential time: 2.22 seconds


### Version 2: Multithreaded (concurrent.futures)

In [30]:
import requests
import time
from concurrent.futures import ThreadPoolExecutor

urls = [
    "https://www.google.com/",
    "https://www.youtube.com/",
    "https://www.python.org",
    "https://www.wikipedia.org",
    "https://www.github.com"
]

def fetch_url(url):
    response = requests.get(url)
    return url, response.status_code

def threaded_download():
    start = time.time()
    with ThreadPoolExecutor(max_workers=5) as executor:
        results = list(executor.map(fetch_url, urls))
    duration = time.time() - start
    for url, status in results:
        print(f"{url} -> {status}")
    print(f"Multithreaded time: {duration:.2f} seconds")

if __name__ == "__main__":
    threaded_download()


https://www.google.com/ -> 200
https://www.youtube.com/ -> 200
https://www.python.org -> 200
https://www.wikipedia.org -> 200
https://www.github.com -> 200
Multithreaded time: 0.77 seconds


### Version 3:  Asynchronous (asyncio + aiohttp)

In [33]:
import asyncio
import aiohttp
import time

urls = [
    "https://www.google.com/",
    "https://www.youtube.com/",
    "https://www.python.org",
    "https://www.wikipedia.org",
    "https://www.github.com"
]

async def fetch_url(session, url):
    async with session.get(url) as response:
        return url, response.status

async def async_download():
    start = time.time()
    async with aiohttp.ClientSession() as session:
        tasks = [fetch_url(session, url) for url in urls]
        results = await asyncio.gather(*tasks)
    duration = time.time() - start
    for url, status in results:
        print(f"{url} -> {status}")
    print(f"Async time: {duration:.2f} seconds")

if __name__ == "__main__":
    await async_download()




https://www.google.com/ -> 200
https://www.youtube.com/ -> 200
https://www.python.org -> 200
https://www.wikipedia.org -> 200
https://www.github.com -> 200
Async time: 1.68 seconds


#### Use timeit to compare execution time

In [36]:
import requests
import aiohttp
import asyncio
import timeit
from concurrent.futures import ThreadPoolExecutor
import nest_asyncio

nest_asyncio.apply()  # Fix for running in Jupyter or notebooks

urls = [
    "https://www.google.com/",
    "https://www.youtube.com/",
    "https://www.python.org",
    "https://www.wikipedia.org",
    "https://www.github.com"
]

# Sequential version
def download_sequential():
    for url in urls:
        response = requests.get(url)
        print(f"Sequential: {url} -> {response.status_code}")

# Multithreaded version
def fetch_threaded(url):
    response = requests.get(url)
    return url, response.status_code

def download_threaded():
    with ThreadPoolExecutor() as executor:
        results = executor.map(fetch_threaded, urls)
        for url, status in results:
            print(f"Threaded: {url} -> {status}")

# Asynchronous version
async def fetch_async(session, url):
    async with session.get(url) as response:
        return url, response.status

async def download_async():
    async with aiohttp.ClientSession() as session:
        tasks = [fetch_async(session, url) for url in urls]
        results = await asyncio.gather(*tasks)
        for url, status in results:
            print(f"Async: {url} -> {status}")

# Function to run async in a sync context
def download_async_wrapper():
    loop = asyncio.get_event_loop()
    loop.run_until_complete(download_async())

# --- Timeit Testing ---
print("\n--- Timing Comparison ---")

seq_time = timeit.timeit(download_sequential, number=1)
print(f"\nSequential Time: {seq_time:.2f} seconds")

threaded_time = timeit.timeit(download_threaded, number=1)
print(f"Threaded Time: {threaded_time:.2f} seconds")

async_time = timeit.timeit(download_async_wrapper, number=1)
print(f"Async Time: {async_time:.2f} seconds")



--- Timing Comparison ---
Sequential: https://www.google.com/ -> 200
Sequential: https://www.youtube.com/ -> 200
Sequential: https://www.python.org -> 200
Sequential: https://www.wikipedia.org -> 200
Sequential: https://www.github.com -> 200

Sequential Time: 2.11 seconds
Threaded: https://www.google.com/ -> 200
Threaded: https://www.youtube.com/ -> 200
Threaded: https://www.python.org -> 200
Threaded: https://www.wikipedia.org -> 200
Threaded: https://www.github.com -> 200
Threaded Time: 0.78 seconds
Async: https://www.google.com/ -> 200
Async: https://www.youtube.com/ -> 200
Async: https://www.python.org -> 200
Async: https://www.wikipedia.org -> 200
Async: https://www.github.com -> 200
Async Time: 0.60 seconds


## Time and Memory usage Comparison
Three version of python script that download HTML content from 5 URLs i.e Sequential, multithreading, & Asynchronous version comparison on the basis of time and memory usage 
- Sequential Version: The implementation with requests library functions by downloading URLs one by one through a sequence.  The program stops running for each request since the previous request must complete which makes it both the easiest and slowest option available.  The execution process takes lengthiest because each URL analysis occurs in sequential order without sharing memory resources.
- Multithreading Version: The ThreadPoolExecutor mechanism enables the multithreaded system to download URLs at the same time.  The process becomes faster in the parallel form when multiple threads simultaneously process requests at the same time.  The high thread operation costs with big URL processing renders it a memory-intensive task.
- Asynchronous Version: This version demands minimal both time and memory resources because it integrates the combination of asyncio and aiohttp.  The asynchronous mode allows simultaneous URL processing without delay because it enables I/O operations to execute in parallel.  The execution time remains short because this method avoids generating excessive threads or processes while maintaining minimum memory consumption.

**Comparing Timing**:

- Execution time with the sequential version becomes the longest because this method proceeds with each request one after another.
- The parallel processing of multiple requests accelerates performance in the multithreaded version yet thread management needs slow down execution time compared to asynchronous methods.
- The asynchronous version proves to be the fastest solution because it manages I/O-bound operations effectively through non-blocking calls.
The asynchronous version outperforms the other resource download methods because it optimizes the execution period and system memory especially in I/O-intensive operations.

##### Use memory_profiler to observe memory usage

In [38]:
import requests
import aiohttp
import asyncio
import timeit
from concurrent.futures import ThreadPoolExecutor
from memory_profiler import profile
import nest_asyncio

nest_asyncio.apply()  # Fix for running in Jupyter or notebooks

urls = [
    "https://www.google.com/",
    "https://www.youtube.com/",
    "https://www.python.org",
    "https://www.wikipedia.org",
    "https://www.github.com"
]

# Sequential version with memory profiling
@profile
def download_sequential():
    for url in urls:
        response = requests.get(url)
        print(f"Sequential: {url} -> {response.status_code}")

# Multithreaded version with memory profiling
@profile
def fetch_threaded(url):
    response = requests.get(url)
    return url, response.status_code

@profile
def download_threaded():
    with ThreadPoolExecutor() as executor:
        results = executor.map(fetch_threaded, urls)
        for url, status in results:
            print(f"Threaded: {url} -> {status}")

# Asynchronous version with memory profiling
@profile
async def fetch_async(session, url):
    async with session.get(url) as response:
        return url, response.status

@profile
async def download_async():
    async with aiohttp.ClientSession() as session:
        tasks = [fetch_async(session, url) for url in urls]
        results = await asyncio.gather(*tasks)
        for url, status in results:
            print(f"Async: {url} -> {status}")

# Function to run async in a sync context with memory profiling
@profile
def download_async_wrapper():
    loop = asyncio.get_event_loop()
    loop.run_until_complete(download_async())

# --- Timeit Testing ---
print("\n--- Timing and Memory Usage Comparison ---")

seq_time = timeit.timeit(download_sequential, number=1)
print(f"\nSequential Time: {seq_time:.2f} seconds")

threaded_time = timeit.timeit(download_threaded, number=1)
print(f"Threaded Time: {threaded_time:.2f} seconds")

async_time = timeit.timeit(download_async_wrapper, number=1)
print(f"Async Time: {async_time:.2f} seconds")



--- Timing and Memory Usage Comparison ---
ERROR: Could not find file C:\Users\sandh\AppData\Local\Temp\ipykernel_13672\4137303280.py
Sequential: https://www.google.com/ -> 200
Sequential: https://www.youtube.com/ -> 200
Sequential: https://www.python.org -> 200
Sequential: https://www.wikipedia.org -> 200
Sequential: https://www.github.com -> 200

Sequential Time: 2.26 seconds
ERROR: Could not find file C:\Users\sandh\AppData\Local\Temp\ipykernel_13672\4137303280.py
ERROR: Could not find file C:\Users\sandh\AppData\Local\Temp\ipykernel_13672\4137303280.py
ERROR: Could not find file C:\Users\sandh\AppData\Local\Temp\ipykernel_13672\4137303280.py
ERROR: Could not find file C:\Users\sandh\AppData\Local\Temp\ipykernel_13672\4137303280.py
ERROR: Could not find file C:\Users\sandh\AppData\Local\Temp\ipykernel_13672\4137303280.py
ERROR: Could not find file C:\Users\sandh\AppData\Local\Temp\ipykernel_13672\4137303280.py
Threaded: https://www.google.com/ -> 200
Threaded: https://www.youtube.c

## Time and Memory Usage by memory-profiler
1. Sequential Version
The first version operates sequentially by getting files individually until every request completes before it starts another.  This approach consumes limited memory while being easy to operate because it deals with requests one by one.  The blocked request system decreases performance levels in comparison to alternative methods.

- Time: This method examines each URL independently which makes it the slowest download process in terms of time duration.

- Memory: Minimal resources get allocated since the protocol manages one request while rejecting concurrent task processing.

2. MultiThreading Version
In the multithreaded version of the program the ThreadPoolExecutor component enables parallel downloading of numerous URLs.  The download process becomes faster when multiple requests operate in parallel threads when compared to the sequential execution method.  The handling of multiple threads for better efficiency results in higher memory demands from managing these threads.

- Time: This version requires slightly longer time than asynchronous processing though it is swifter than the sequential version thanks to its capability to process activities concurrently.

- Memory: The simultaneous use of memory by individual threads within this version makes it require greater memory than the single-threaded sequential implementation.

3. Asynchronous Version
The asynchronous version oversees multiple simultaneous requests through asyncio and aiohttp which prevents blocking delays.  Memory utilization stays low thus the software can handle numerous requests effectively.  The asynchronous method operates at maximum speed because it skips the requirement for request completion before beginning new ones.

- Time:The asynchronous version manages multiple requests simultaneously which makes it operate at the fastest speed among the three versions.

- Memory:This application requires minimal memory because the asynchronous tasks it uses require less memory than thread processes.

## **Pros and cons of each version**
1. Advantages of Sequential Version:

- The application's concurrency features and thread management remain simple and intuitive to implement.

- Very little memory is required since the system handles only one request at a time.

- Such configuration works best for applications managing a reduced number of URLs and requests  (Sequential Vs Asynchronous Programming in Python – SemFio Networks, n.d.)

- Cons:

- şerential request processing becomes slow when dealing with large or multiple URLs and extensive datasets.

- The blocking design of the program wastes processing time because it needs to wait until the first request completes before starting the next one.

2. Advantages of Multithreaded Version:

- The utilization of multiple threads within this version enables faster operation as it can download multiple URLs simultaneously better than the sequential version.

- The design works well with activities that contain an average number of requests but no CPU performance constraints (What Are the Pros and Cons of Multithreading?, n.d.)



- Cons:

- Memory usage increases because each created thread requires individual memory storage.

- The management structure of threads consumes additional overhead because it requires multiple resources to create and track threads.

- Using this technique is inappropriate for big execution tasks since it can reduce performance when there are numerous threads or insufficient system resources(What Are the Pros and Cons of Multithreading?, n.d.)


3. Advantages of the Asynchronous Version:

- Such a design delivers maximum speed with its asynchronous functionality that allows concurrent request execution through non-blocking operations.

- Asynchronous modules split resource allocation between them while using single threads so they occupy less memory space than threading systems.

- The asynchronous version functions best for managing many requests with its efficiency on input/output workloads such as web scraping applications (Sequential Vs Asynchronous Programming in Python – SemFio Networks, n.d.)

- Cons:

- For beginners the difficulty of understanding and fixing asynchronous programs increases which makes them more complicated to learn.

- aiohttp and asyncio provide force code execution through additional libraries which might require extra management and setup according to applications.

- The technology provides minimal assistance to CPU-intensive tasks but performs exceptionally well on jobs that depend on I/O operations.

### Compare and discuss which approach is most efficient and why

- Asynchronous technique represents the most effective solution for I/O-bound operations including URL downloads because it enables simultaneous processing without blocking operations.  The system executes multiple tasks simultaneously thus reducing execution time since it does not need to wait for previous tasks to complete.  Only one thread operates in asynchronous programming while multithreaded systems requires considerable memory utilization.  A large number of I/O-bound operations can be best managed through this option due to its high scalability and ability to work with many running tasks simultaneously.
- The multithreaded technique outperforms sequential processing because it enables many threads to work on I/O tasks at the same time.  The system incurs additional memory overhead due to thread management that becomes less efficient as the number of concurrent operations increases.  The asynchronous programming method offers better time-to-scale ratio than sequencing yet performs better at larger levels of concurrent operations.
- The sequential processing procedure provides the least difficulty yet achieves the worst execution efficiency during multiple I/O operations.  Magnifying the number of URLs for download increases the required time due to its single-request approach to handle each request.  The low memory utilization of this method makes it incompatible for parallel request processing since it lacks concurrent execution capabilities.
- The most efficient processing method for managing many I/O-bound operations like URL downloads appears to be asynchronous programming followed by multithreaded approaches for balancing concurrency requirements and the sequential approach as the most basic yet resource-efficient solution.

**Reference**
- Sequential vs Asynchronous programming in Python – SemFio Networks. (n.d.). https://semfionetworks.com/blog/sequential-vs-asynchronous-programming-in-python/
- What are the pros and cons of multithreading? (n.d.). Tech Interview Preparation – System Design, Coding & Behavioral Courses | Design Gurus. https://www.designgurus.io/answers/detail/what-are-the-pros-and-cons-of-multithreading
